evm/src/arithmetic/mod.rs

```diff
@@ -143,19 +130,28 @@ pub(crate) enum Operation {
 }
 
 impl Operation {
-    pub(crate) fn binary(
-        operator: BinaryOperator,
-        input0: U256,
-        input1: U256,
-        is_simulated: bool,
-    ) -> Self {
+    /// Create a binary operator with given inputs.
+    ///
+    /// NB: This works as you would expect, EXCEPT for SHL and SHR,
+    /// whose inputs need a small amount of preprocessing. Specifically,
+    /// to create `SHL(shift, value)`, call (note the reversal of
+    /// argument order):
+    ///
+    ///    `Operation::binary(BinaryOperator::Shl, value, 1 << shift)`
```

Comment:
At the very beginning of this flag combination work, I thought of something along those lines, but then reverted it because I found it a bit confusing (calling a shift operation with an already shifted operand), though I have to admit this is overall much cleaner than my approach, and the comments should be self-explanatory.